include/glow/Support/Register.h

```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_SUPPORT_REGISTER_H
+#define GLOW_SUPPORT_REGISTER_H
+
+#include <cassert>
+#include <unordered_map>
+
+namespace glow {
+
+template <class Key, class Base> class BaseFactory {
+public:
+  virtual ~BaseFactory() = default;
+  virtual Base *create() = 0;
+  virtual Key getRegistrationKey() const = 0;
+};
+
+template <class Key, class Base> class FactoryRegistry {
+public:
+  using FactoryMap = std::unordered_map<Key, BaseFactory<Key, Base> *>;
+
+  static void registerFactory(BaseFactory<Key, Base> &factory) {
+    Key registrationKey = factory.getRegistrationKey();
+    auto inserted = factories().emplace(registrationKey, &factory);
+    assert(inserted.second && "Double registration of base factory");
+  }
+
+  static Base *get(const Key &key) {
+    auto it = factories().find(key);
+
+    if (it == factories().end()) {
+      return nullptr;
+    }
+
+    return it->second->create();
+  }
+
+private:
+  static FactoryMap &factories() {
+    static FactoryMap *factories = new FactoryMap();
```

Comment:
Does asan complain about this?

Reply:
that should be fine, see here: https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use

Reply:
Cool thanks, I also found this piece from there interesting and it also touches on what I was saying https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2.